### PR TITLE
Refactor hash_labels for family.

### PR DESCRIPTION
**Pull request by jovan-wong**
This PR aims to refactor the hash_labels for family with our defined hash_combine:
1. Add `hash_combine` and `hash_value` functions.
2. Refactor `hash_labels`.
3. Add a simple ut.

some issues in need of help:
1. I don't know what's the best practice to test a function of computing hash value.
2. I'm not sure whether it's good to add a utils file and put `hash_combine` and `hash_value` to it.

And any idea for this?

### CODE REVIEW

**Comment by jovan-wong** [comment]
@gjasny  @jupp0r cc

**Comment by jovan-wong** [comment]
@gjasny  CI failed. I think it's not related my changes.

something wrong with ci?

**Comment by gjasny** [comment]
I'll take a look at CI

**Comment by gjasny** [comment]
Works for master commit, again: https://travis-ci.org/jupp0r/prometheus-cpp/builds/463523040?utm_source=github_status&utm_medium=notification

File: core/include/prometheus/detail/hash.h [hunk 1]
```diff
@@ -0,0 +1,46 @@
+#pragma
+
+#include <functional>
+
+namespace prometheus {
+
```
Comment (gjasny) [inline core/include/prometheus/detail/hash.h]: Could you please add `namespace detail {` here as well?

File: core/include/prometheus/detail/utils.h [hunk 2]
```diff
@@ -0,0 +1,20 @@
+#pragma onece
+
+#include <map>
+#include <cstddef>
+#include <string>
+
+namespace prometheus {
+
+namespace utils {
```
Comment (gjasny) [inline core/include/prometheus/detail/utils.h]: I'd rename that namespace to `detail`

File: core/include/prometheus/detail/hash.h [hunk 3]
```diff
@@ -0,0 +1,46 @@
+#pragma
+
+#include <functional>
```
Comment (gjasny) [inline core/include/prometheus/detail/hash.h]: You also need a `<cstddef>` header here for `std::size_t`

File: core/include/prometheus/detail/hash.h [hunk 4]
```diff
@@ -0,0 +1,46 @@
+#pragma
```
Comment (gjasny) [inline core/include/prometheus/detail/hash.h]: There's no need for that file to be a public header. Please move it to `core/src/detail/`.

File: core/tests/utils_test.cc [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+  EXPECT_EQ(value1, value2);
+}
+
+
+TEST(UtilsTest, hash_labels_2) {
+  std::map<std::string, std::string> labels1;
+  labels1.insert(std::make_pair<std::string, std::string>("aa", "bb"));
+  std::map<std::string, std::string> labels2;
+  labels2.insert(std::make_pair<std::string, std::string>("a", "abb"));
+  expect_not_equal(labels1, labels2);
```
Comment (gjasny) [inline core/tests/utils_test.cc]: I'd inline it into `EXPECT_NE(utils::hash_labels(labels1), utils::hash_labels(labels2))`

File: core/tests/utils_test.cc [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+  std::map<std::string, std::string> labels3;
```
Comment (gjasny) [inline core/tests/utils_test.cc]: and split into two functions

File: core/tests/utils_test.cc [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+  labels.insert(std::make_pair<std::string, std::string>("key2", "vaule2"));
+  auto value1 = utils::hash_labels(labels);
+  auto value2 = utils::hash_labels(labels);
+
+  EXPECT_EQ(value1, value2);
+}
+
+
+TEST(UtilsTest, hash_labels_2) {
+  std::map<std::string, std::string> labels1;
```
Comment (gjasny) [inline core/tests/utils_test.cc]: In C++11 you could also write it like `std::map<std::string, std::string> labels1{{"aa", "bb"}};`. That has the advantage that you could make the var const if you want.

File: core/include/prometheus/detail/hash.h (hunk 4 above)
Comment (jovan-wong) [reply core/include/prometheus/detail/hash.h]: I could not understand that why we move `hash.h` file into `core/src/detail` folder.

If we compile this project to a static/shared library,  and provide `core/include` for users, how can users include `hash.h` file?

**Comment by jovan-wong** [comment]
@gjasny  I have addressed all the comments.
Could u plz take a look again?

**Comment by jovan-wong** [comment]
@gjasny  I have addressed all the comments.
Could u plz take a look again?

File: core/include/prometheus/detail/hash.h (hunk 4 above)
Comment (gjasny) [reply core/include/prometheus/detail/hash.h]: Right now users don't have to see that file so we could hide it as an implementation detail and won't have to take care of API / ABI compatibility.

**Comment by gjasny** [comment]
Thanks a lot for your work!

**Pull request merged**